File: java/src/main/java/mchehab/com/java/MainActivity.java
package mchehab.com.java;

import android.content.Intent;
import android.support.design.widget.FloatingActionButton;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;

import java.util.ArrayList;
import java.util.List;

import mchehab.com.java.click_listeners.OnDeleteListener;
import mchehab.com.java.click_listeners.OnEditListener;

public class MainActivity extends AppCompatActivity implements OnEditListener, OnDeleteListener{

    private FloatingActionButton floatingActionButton;
    private RecyclerView recyclerView;

    private PersonAdapter personAdapter;
    private List<Person> listPerson;

    private final int REQUEST_CODE_EDIT = 101;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        listPerson = new ArrayList<>();
        for(int i=0;i<10;i++){
            listPerson.add(new Person("F" + (i+1), "L" + (i+1)));
        }

        floatingActionButton = findViewById(R.id.floatingActionButton);
        recyclerView = findViewById(R.id.recyclerView);

        personAdapter = new PersonAdapter(listPerson, this, this);
        recyclerView.setAdapter(personAdapter);
        recyclerView.setLayoutManager(new LinearLayoutManager(this, LinearLayoutManager.VERTICAL,
         false));

        setFloatingActionButtonListener();
    }

    private void setFloatingActionButtonListener() {
        floatingActionButton.setOnClickListener(e -> {
            Intent intent = new Intent(this, EditActivity.class);
            intent.putExtra(Constants.PERSON_INTENT_EDIT, false);
            startActivityForResult(intent, REQUEST_CODE_EDIT);
        });
    }

    @Override
    public void deleteItem(Person person) {
        listPerson.remove(person);
        personAdapter.notifyDataSetChanged();
    }

    @Override
    public void editItem(Person person, int index) {
        Intent intent = new Intent(this, EditActivity.class);
        intent.putExtra(Constants.PERSON_INTENT_EDIT, true);
        intent.putExtra(Constants.PERSON_INTENT_INDEX, index);
        intent.putExtra(Constants.PERSON_INTENT_OBJECT, person);
        startActivityForResult(intent, REQUEST_CODE_EDIT);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if(requestCode == REQUEST_CODE_EDIT){
            if(resultCode == RESULT_OK){
                if(data == null){
                    return;
                }
                boolean isEdit = data.getBooleanExtra(Constants.PERSON_INTENT_EDIT, false);
                Person person = data.getParcelableExtra(Constants.PERSON_INTENT_OBJECT);
                if(isEdit){
                    int index = data.getIntExtra(Constants.PERSON_INTENT_INDEX, -1);
                    if(index == -1){
                        return;
                    }
                    listPerson.set(index, person);
                    personAdapter.notifyDataSetChanged();
                }else{
                    listPerson.add(person);
                    personAdapter.notifyDataSetChanged();
                }
            }
        }
    }
}

File: app/src/main/java/mchehab/com/advancedrecyclerview/MainActivity.kt
package mchehab.com.advancedrecyclerview

import android.app.Activity
import android.content.Intent
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.support.v7.widget.LinearLayoutManager
import kotlinx.android.synthetic.main.activity_main.*
import mchehab.com.advancedrecyclerview.click_listeners.OnDeleteListener
import mchehab.com.advancedrecyclerview.click_listeners.OnEditListener

class MainActivity : AppCompatActivity(), OnDeleteListener, OnEditListener {

    private val REQUEST_CODE = 101

    private lateinit var personAdapter: PersonAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        val listPerson = mutableListOf(Person("f1", "f2"),
                Person("f2", "l2"),
                Person("f3", "l3"),
                Person("F4", "L4"),
                Person("f5", "l5"),
                Person("f6", "l6"),
                Person("F7", "L7"),
                Person("f8", "l8"),
                Person("f9", "l9"),
                Person("F10", "L10"))

        personAdapter = PersonAdapter(listPerson, this, this)
        recyclerView.adapter = personAdapter
        recyclerView.layoutManager = LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false)

        setFloatingActionButtonListener()
    }

    private fun setFloatingActionButtonListener(){
        floatingActionButton.setOnClickListener {
            val intent = Intent(this@MainActivity, EditActivity::class.java)
            intent.putExtra(PERSON_INTENT_EDIT, false)
            startActivityForResult(intent, REQUEST_CODE)
        }
    }

    override fun deleteItem(person: Person) {
        personAdapter.listPerson.remove(person)
        personAdapter.notifyDataSetChanged()
    }

    override fun editItem(person: Person, index: Int) {
        val intent = Intent(this@MainActivity, EditActivity::class.java)
        intent.putExtra(PERSON_INTENT_EDIT, true)
        intent.putExtra(PERSON_INTENT_OBJECT, person)
        intent.putExtra(PERSON_INTENT_INDEX, index)
        startActivityForResult(intent, REQUEST_CODE)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (requestCode == REQUEST_CODE){
            if(resultCode == Activity.RESULT_OK && data != null){
                val bundle = data.extras
                val isEdit = bundle.getBoolean(PERSON_INTENT_EDIT, false)
                val person = bundle.getParcelable<Person>(PERSON_INTENT_OBJECT)
                if(isEdit){
                    val index = bundle.getInt(PERSON_INTENT_INDEX, -1)
                    // something very wrong occurred, do nothing
                    if (index == -1)
                        return
                    personAdapter.listPerson.set(index, person)
                    personAdapter.notifyItemChanged(index)
                }else{
                    personAdapter.listPerson.add(person)
                    personAdapter.notifyItemInserted(personAdapter.listPerson.size-1)
                }
            }
        }
    }
}